028-service-binding.md

```diff
@@ -0,0 +1,420 @@
+      listeners: aW50ZXJuYWxfa2Fma2Euc3ZjOjkwOTIsZXh0ZXJuYWxfbXlob3N0LmNvbTo0NDM= # when base64 decoded something like internal_kafka.svc:9092,external_myhost.com:443
+      ca.crt: # Strimzi cluster CA certificate
+      ca.p12: # PKCS #12 archive file for Strimzi cluster CA certificate
+      ca.password: # Password for protecting the Strimzi cluster CA certificate PKCS #12 archive file
+    ```
+  - e.g. the secret could contain a separate entry for each listener:
+    ```yaml
+    data:
+    listener.internal: aW50ZXJuYWxfbG9jYWxob3N0OjkwODA= # when base64 decoded something like internal_kafka.svc:9092
+    listener.external: ZXh0ZXJuYWxfbXlob3N0LmNvbTo0NDM= # when base64 decoded something like external_myhost.com:443
+    ca.crt: # Strimzi cluster CA certificate
+    ca.p12: # PKCS #12 archive file for Strimzi cluster CA certificate
+    ca.password: # Password for protecting the Strimzi cluster CA certificate PKCS #12 archive file
+    ```
+    
+**Note:** The Service Binding spec does contain some suggested secret fields for certificates, but they will not suffice to encapsulate all the certificate related information that an application needs. Hence, the suggestion of the separate fields above.
+
+Considerations:
+
+ - This approach results in a new secret being required that contains copies of existing secrets
+ - It does not help the user to decide which listener to use. The application would need to parse the secret to pick out the correct listener from the list
+
+### 3 - Application binds to KafkaUser resource that contains all details
+
+In this approach the application only binds to the `KafkaUser` and the secret created for the `KafkaUser` contains all the required details.
+
+Required changes:
+
+- Update the `KafkaUser` CR to include a `status.binding` field where the `name` matches the current field `secret` in the status
+- Update the secret created for the `KafkaUser` to include the username when the `KafkaUser` is using type SASL SCRAM
+- Update the secret created for the `KafkaUser` to include the bootstrap servers for each listener
+- Update the secret created for the `KafkaUser` to include the cluster certificate information
+
+Considerations:
+
+- This approach results in a lot more information being added to each `KafkaUser` secret
+- It does not help the user to decide which listener to use. The application would need to parse the secret to pick out the correct listener from the list
+
+Possible extensions to allow only one bootstrap server to be listed:
+
+ - Could only add bootstrap server addresses for listeners that support the authentication type that the `KafkaUser` supports
+ - Could update the `KafkaUser` CR to include `spec.binding.listener` field which would then determine the listener bootstrap server address that is added to the secret
+
+### 4 - Application can bind to a single custom resource
+
+In this approach the application binds to a new custom resource type, e.g. called `KafkaConnection`.
```

Comment:
Share Paolo's concern on name being confused with Kafka Connect. Is something like `KafkaServiceBinding` feasible?

Reply:
Well, as far as I understood it:
* This is not a service binding => this is a resource the service binding will bind to
* The Secret this creates is usable outside of Service binding AFAIU => so we should not tie the name to it

I'm not sure I see the confusion => while the names are similar, they will be used in different contexts a possibly by different personas.

Reply:
I totally agree that the `Secret` related to this binding resource is actually usable outside of the service binding and it's really useful to users; this could even be generated out of the needs for the service binding operator.
The `KafkaConnection` resource is instead a resource that makes sense only in the service binding operator land, or?
As from the proposal:

> The Strimzi operator adds a `status.binding` to the `KafkaConnection` status that points to a Kubernetes `Secret`

So the `KafkaConnection` is really something related to the support for service binding operator. 
Then, on the service binding specification when they mention the `ServiceBinding` resource they have two main info "workload" (the actual user application) and the "service" (in our case Kafka).
Would it make more sense to call use `KafkaService` instead?

Reply:
> The KafkaConnection resource is instead a resource that makes sense only in the service binding operator land, or?

You need the `KafkaConnection` to get the Secret. So they are tied together. So if you wanna use the secret without the Service Binding operator, you still need to use the KafkaConnection resource => you create the KafkaConnection resource and the KafkaConnection operator will create the secret.

> Would it make more sense to call use `KafkaService` instead?

I think that this is more confusing than `KafkaConnection` -> it mixes with Kubernetes `Service`, so I would expect that to create some endpoint and not just a secret. Although I guess this could be individual.

Reply:
> You need the KafkaConnection to get the Secret. So they are tied together.

They are tied together from a service binding operator point of view. It's the KafkaConnection referring to the Secret via the `service.binding` field not the other way around. So if you want to use the Secret alone you don't need the KafkaConnection.

Maybe it's more related to how we envisage the implementation as you were mentioning. It's a new operator creating both KafkaConnection and Secret, starting from the CO resources, right? Why you said something like "you create the KafkaConnection resource and the KafkaConnection operator will create the secret." ?

Ragarding the name I think that `KafkaService` is really more consistent with the Service Binding, their spec and their examples and it's really just related to it.

Reply:
@ppatierno @scholzj @tombentley any more thoughts on the naming of the new CR?

Reply:
My personal feeling is that `KafkaService` fits better in the service binding picture, but it's just my opinion of course.

Reply:
I'm also inclined to `KafkaService`, since it's not really a "connection" in a meaningful sense (sure, the information necessary to make a connection, but that's not the same thing), and it's for _service_ binding.

Reply:
Its not a service either ;-). I think `KafkaConnection` better represents what do you use it for.

Reply:
And again ... this is not really depending in any way on the Service Binding. The `Kafkaconnection` or whatever the name is super useful on its own without the Service Binding => it gives you the secret and you can just mount it and user it without the Service Binding operator.

Reply:
I prefer `KafkaService` over `KafkaConnection` because if I was looking for the address of where my Kafka was in Kubernetes I would look for a `Service`. Going in a different direction though we could go with `KafkaBindingRequest`, as it's not just the Service Binding spec that uses binding to mean tying an application to a service, and you are making a request to get the information you need to bind to the Kafka.

Reply:
I still prefer `KafkaService` ...

Reply:
Throwing some other ideas down.. `KafkaAccess`, `KafkaApplication`, `KafkaTenant`, `KafkaCredentials`, `KafkaPortal`, and the two I like but we probably can't use due to their use elsewhere `KafkaIngress`, `KafkaGateway`

Reply:
I still feel like `KafkaService` and `KafkaConnection` are probably better than any of the other ones I've come up with though

Reply:
What about `KafkaBinding`?

Reply:
> I prefer KafkaService over KafkaConnection because if I was looking for the address of where my Kafka was in Kubernetes I would look for a Service.

But it is not service -> that has consequences on access rights, for network policies and tenant segregation etc. If you have a Kube service, you can connect to it. But this resource has nothing to do with service.

> Throwing some other ideas down.. KafkaAccess, KafkaApplication, KafkaTenant, KafkaCredentials, KafkaPortal, and the two I like but we probably can't use due to their use elsewhere KafkaIngress, KafkaGateway

I think `KafkaAccess` is good -> no conflicts with native Kube resources and it fairly well describes what you can get ... information to access the Kafka cluster.

`KafkaApplication` is probably something I can live with. But I would really expect that you pass some DSL inside the resource and it builds the Kafka application for you ;-). So not sure how understandable it is.

`KafkaTenant`, `KafkaPortal` and `KafkaCredentials` IMHO don't describe what it is and how is it used. `KafkaIngress` and `KafkaGateway` bring the same confusion as `KafkaService` where pre-existing resource create expectations of what it does.

Reply:
Sorry I am still with `KafkaService` which matches what service binding is and what I saw in the spec ... but I don't want to force on this, maybe we should start a voting process :P

Reply:
Discussed on the community call on 2nd Dec -> we should go with `KafkaAccess`.